File: backend/app/api/routes/customers.py
"""
Customer Routes
===============
Authenticated customer read operations.

Endpoints:
    GET /api/customers          — list customers (search + pagination) (auth)
    GET /api/customers/{id}     — get one customer + order history (auth)

Write operations (create/update) are future work and are intentionally NOT
implemented in Phase 4A.
"""

import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_db
from app.models.admin import AdminUser
from app.schemas.common import Page, PageParams
from app.schemas.customer import CustomerDetailRead, CustomerRead
from app.services import customer_service

router = APIRouter(prefix="/api/customers", tags=["Customers"])


@router.get(
    "",
    response_model=Page[CustomerRead],
    summary="List customers",
)
async def list_customers(
    params: PageParams = Depends(),
    search: str | None = Query(default=None, description="Search by name, phone, or email"),
    db: AsyncSession = Depends(get_db),
    _: AdminUser = Depends(get_current_user),
) -> dict:
    """
    List customers with optional search and pagination.

    Query params:
      - page, page_size   — pagination
      - search            — case-insensitive match against name/phone/email
    """
    customers, total = await customer_service.list_customers(
        db,
        page=params.page,
        page_size=params.page_size,
        search=search,
    )
    pages = (total + params.page_size - 1) // params.page_size
    return {
        "items": customers,
        "total": total,
        "page": params.page,
        "page_size": params.page_size,
        "pages": pages,
    }


@router.get(
    "/{customer_id}",
    response_model=CustomerDetailRead,
    summary="Get a customer with their order history",
)
async def get_customer(
    customer_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: AdminUser = Depends(get_current_user),
) -> CustomerDetailRead:
    """
    Return a single customer including their order history (newest first).

    Raises HTTP 404 if the customer does not exist.
    """
    return await customer_service.get_customer_with_orders(db, customer_id)

File: src/lib/api/customers.ts
/**
 * Customers API
 * =============
 * Typed functions for the Phase 4A customer endpoints, using the centralized
 * API client so the auth token is attached automatically.
 *
 * Endpoints consumed:
 *   GET /api/customers          — list customers (search + pagination, auth)
 *   GET /api/customers/{id}     — one customer + order history (auth)
 *
 * Write ops (create/update customers) are future work on the backend and are
 * intentionally not exposed here.
 */

import { apiClient } from "@/lib/api/client";
import type {
  CustomerDetailRecord,
  CustomerListParams,
  CustomerRecord,
  PageResult,
} from "@/lib/types";

function toQuery(params: object): string {
  const query = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (
      value !== undefined &&
      (typeof value === "string" || typeof value === "number")
    ) {
      query.set(key, String(value));
    }
  }
  const qs = query.toString();
  return qs ? `?${qs}` : "";
}

/** List customers (most recently updated first) with optional search + pagination. */
export async function getCustomers(
  params: CustomerListParams = {}
): Promise<PageResult<CustomerRecord>> {
  return apiClient.get<PageResult<CustomerRecord>>(
    `/api/customers${toQuery(params)}`
  );
}

/** Get one customer including their order history (newest first). */
export async function getCustomer(
  customerId: string
): Promise<CustomerDetailRecord> {
  return apiClient.get<CustomerDetailRecord>(`/api/customers/${customerId}`);
}